cmd/proxy/wsl_integration_linux.go

```diff
@@ -56,14 +74,50 @@ func main() {
 	ctx, _ := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
 	proxy := httputil.NewSingleHostReverseProxy(targetURL)
 
+	certPool, err := x509.SystemCertPool()
+	if err != nil {
+		logrus.Fatalf("failed loading System Cert Pool: %v", err)
+	}
+
+	crtByte, err := os.ReadFile(rootCA)
+	if err != nil {
+		logrus.Fatalf("failed reading upstream CA Cert: %v", err)
+	}
+
+	ok := certPool.AppendCertsFromPEM(crtByte)
+	if !ok {
+		logrus.Fatal("failed to append Root CA PEM")
+	}
+
+	upstreamKeyPair, err := tls.LoadX509KeyPair(upstreamCertFile, upstreamKeyFile)
+	if err != nil {
+		logrus.Fatalf("failed to load upstream certificate/key: %v", err)
+	}
+	proxy.Transport = &http.Transport{
+		TLSClientConfig: &tls.Config{ // #nosec G402 the min TLS version is acceptable
+			Certificates: []tls.Certificate{upstreamKeyPair},
+			RootCAs:      certPool,
+		},
+	}
+
 	srv := http.Server{
 		Addr:              listenAddr,
 		Handler:           proxy,
 		ReadHeaderTimeout: 5 * time.Second,
 	}
 
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		logrus.Fatalf("failed to load downstream certificate/key: %v", err)
+	}
+
+	srv.TLSConfig = &tls.Config{ // #nosec G402 the min TLS version is acceptable
+		PreferServerCipherSuites: true,
+		Certificates:             []tls.Certificate{cert},
+	}
+
 	go func() {
-		if err := srv.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+		if err := srv.ListenAndServeTLS("", ""); !errors.Is(err, http.ErrServerClosed) {
```

Comment:
Oh, right, we can avoid loading calling `tls.LoadX509KeyPair()` manually and just pass them in to `ListenAndServeTLS()`…

Reply:
Yes, that makes sense. For some reason, I assumed that only accepts `.pem` format, but that is not the case since the underlying code calls `tls.LoadX509KeyPair()` anyways.